dplutils/pipeline/stream.py

```diff
@@ -136,39 +138,41 @@ def resolve_completed(self):
                 task.data_in.extendleft(self.task_resolve_output(ready))
         return None
 
-    def process_source(self, source):
-        source_batch = []
-        for _ in range(source.task.batch_size or 1):
+    def _feed_source(self, source):
+        if self.source_exhausted:
+            return
+        total_length = sum(i.length for i in source.data_in)
+        while total_length < (source.task.batch_size or 1):
             try:
-                source_batch.append(next(self.source_generator))
+                next_df = next(self.source_generator)
             except StopIteration:
                 self.source_exhausted = True
-                if len(source_batch) == 0:
-                    return
                 break
+            # We feed any generated source to all source tasks similar the way
+            # upstream forked outputs broadcast. We add to data_in so that any
+            # necessary batching and splitting can be handled by normal procedure.
+            for task in self.stream_graph.source_tasks:
+                task.data_in.append(StreamBatch(data=next_df, length=len(next_df)))
             self.n_sourced += 1
             if self.n_sourced == self.max_batches:
                 self.source_exhausted = True
                 break
-        source.pending.appendleft(self.task_submit(source.task, source_batch))
```

Comment:
this diff is a bit confusing I think. The change here makes it so that the generator sourcing uses the same scheduling facility as all other tasks after we have pulled off the generator. So basically the work that this did is not a combination of line 155:

```py
    task.data_in.append(StreamBatch(data=next_df, length=len(next_df)))
```

which uses the data_in queue like all other tasks do, and call to `_handle_one_task` on 211, which ends up making the `pending.appendleft` on 189.